src/test/test-hostname-setup.c

```diff
@@ -3,58 +3,76 @@
         _cleanup_(unlink_tempfilep) char path[] = "/tmp/hostname.XXXXXX";
         char *hostname;
-        int fd;
+        int r;
 
-        fd = mkostemp_safe(path);
-        assert_se(fd > 0);
-        close(fd);
+        safe_close(ASSERT_FD(mkostemp_safe(path)));
 
         /* simple hostname */
-        assert_se(write_string_file(path, "foo", WRITE_STRING_FILE_CREATE) == 0);
-        assert_se(read_etc_hostname(path, &hostname) == 0);
+        ASSERT_OK(write_string_file(path, "foo", WRITE_STRING_FILE_CREATE));
+        ASSERT_OK(read_etc_hostname(path, /* substitute_wildcards= */ false, &hostname));
         ASSERT_STREQ(hostname, "foo");
         hostname = mfree(hostname);
 
         /* with comment */
-        assert_se(write_string_file(path, "# comment\nfoo", WRITE_STRING_FILE_CREATE) == 0);
-        assert_se(read_etc_hostname(path, &hostname) == 0);
-        assert_se(hostname);
+        ASSERT_OK(write_string_file(path, "# comment\nfoo", WRITE_STRING_FILE_CREATE));
+        ASSERT_OK(read_etc_hostname(path, /* substitute_wildcards= */ false, &hostname));
+        ASSERT_NOT_NULL(hostname);
         ASSERT_STREQ(hostname, "foo");
         hostname = mfree(hostname);
 
         /* with comment and extra whitespace */
-        assert_se(write_string_file(path, "# comment\n\n foo ", WRITE_STRING_FILE_CREATE) == 0);
-        assert_se(read_etc_hostname(path, &hostname) == 0);
-        assert_se(hostname);
+        ASSERT_OK(write_string_file(path, "# comment\n\n foo ", WRITE_STRING_FILE_CREATE));
+        ASSERT_OK(read_etc_hostname(path, /* substitute_wildcards= */ false, &hostname));
+        ASSERT_NOT_NULL(hostname);
         ASSERT_STREQ(hostname, "foo");
         hostname = mfree(hostname);
 
         /* cleans up name */
-        assert_se(write_string_file(path, "!foo/bar.com", WRITE_STRING_FILE_CREATE) == 0);
-        assert_se(read_etc_hostname(path, &hostname) == 0);
-        assert_se(hostname);
+        ASSERT_OK(write_string_file(path, "!foo/bar.com", WRITE_STRING_FILE_CREATE));
+        ASSERT_OK(read_etc_hostname(path, /* substitute_wildcards= */ false, &hostname));
+        ASSERT_NOT_NULL(hostname);
         ASSERT_STREQ(hostname, "foobar.com");
         hostname = mfree(hostname);
 
+        /* with wildcards */
+        ASSERT_OK(write_string_file(path, "foo????????x??????????u", WRITE_STRING_FILE_CREATE));
+        ASSERT_OK(read_etc_hostname(path, /* substitute_wildcards= */ false, &hostname));
+        ASSERT_NOT_NULL(hostname);
+        ASSERT_STREQ(hostname, "foo????????x??????????u");
+        hostname = mfree(hostname);
+        r = read_etc_hostname(path, /* substitute_wildcards= */ true, &hostname);
+        if (ERRNO_IS_NEG_MACHINE_ID_UNSET(r))
+                log_tests_skipped("skipping wildcard hostname tests, no machine ID defined");
+        else {
+                ASSERT_OK(r);
+                ASSERT_NOT_NULL(hostname);
+                ASSERT_FALSE(streq(hostname, "foo????????x??????????u"));
```

Comment:
`ASSERT_NULL(strchr(hostname, '?'))` is better?
I'd like to also add `hostname_is_valid()` test for the generated hostname.